Sources/SwiftMemcache/MemcachedResponseDecoder.swift

```diff
@@ -0,0 +1,178 @@
+struct MemcachedResponseDecoder: NIOSingleStepByteToMessageDecoder {
+    typealias InboundOut = MemcachedResponse
+
+    /// Describes the errors that can occur during the decoding process.
+    enum MemcachedDecoderError: Error {
+        /// This error is thrown when EOF is encountered but there are still
+        /// readable bytes in the buffer, which can indicate a bad message.
+        case unexpectedEOF
+
+        /// This error is thrown when EOF is encountered but the decoder's next step
+        /// is not `.none`. This error suggests that the message ended prematurely,
+        /// possibly indicating a bad message.
+        case unexpectedNextStep(NextStep)
+    }
+
+    /// The next step that the decoder will take. The value of this enum determines how the decoder
+    /// processes the current state of the ByteBuffer.
+    enum NextStep: Hashable {
+        /// No further steps are needed, the decoding process is complete for the current message.
+        case none
+        /// The initial step.
+        case returnCode
+        /// Decode the data length, flags or check if we are the end
+        case dataLengthOrFlag(MemcachedResponse.ReturnCode)
+        /// Decode the next flag
+        case decodeNextFlag(MemcachedResponse.ReturnCode, UInt64?)
+        /// Decode end of line
+        case decodeEndOfLine(MemcachedResponse.ReturnCode, UInt64?)
+    }
+
+    /// The action that the decoder will take in response to the current state of the ByteBuffer and the `NextStep`.
+    enum NextDecodeAction {
+        /// We need more bytes to decode the next step.
+        case waitForMoreBytes
+        /// We can continue decoding.
+        case continueDecodeLoop
+        /// We have decoded the next response and need to return it.
+        case returnDecodedResponse(MemcachedResponse)
+    }
+
+    /// The next step in decoding.
+    var nextStep: NextStep = .returnCode
+
+    mutating func decode(buffer: inout ByteBuffer) throws -> InboundOut? {
+        while self.nextStep != .none {
+            switch try self.next(buffer: &buffer) {
+            case .returnDecodedResponse(let response):
+                return response
+
+            case .waitForMoreBytes:
+                return nil
+
+            case .continueDecodeLoop:
+                ()
+            }
+        }
+        return nil
+    }
+
+    mutating func next(buffer: inout ByteBuffer) throws -> NextDecodeAction {
+        switch self.nextStep {
+        case .none:
+            return .waitForMoreBytes
+
+        case .returnCode:
+            guard let bytes = buffer.readInteger(as: UInt16.self) else {
+                return .waitForMoreBytes
+            }
+
+            let returnCode = MemcachedResponse.ReturnCode(bytes)
+            self.nextStep = .dataLengthOrFlag(returnCode)
+            return .continueDecodeLoop
+
+        case .dataLengthOrFlag(let returnCode):
+            if returnCode == .stored {
```

Comment:
This is not correct we should only decode the data length if the returnCode is `VA`